src/file.ts

```diff
@@ -1562,31 +1572,13 @@ class File extends ServiceObject<File> {
           return;
         }
 
-        // TODO(https://github.com/googleapis/nodejs-storage/issues/709):
-        // Remove once the backend issue is fixed.
-        // If object is stored compressed (having
-        // metadata.contentEncoding === 'gzip') and was served decompressed,
-        // then skip checksum validation because the remote checksum is computed
-        // against the compressed version of the object.
-        if (!isServedCompressed) {
-          try {
-            await this.getMetadata({userProject: options.userProject});
-          } catch (e) {
-            throughStream.destroy(e as Error);
-            return;
-          }
-          if (this.metadata.contentEncoding === 'gzip') {
-            throughStream.end();
-            return;
-          }
-        }
-
         // If we're doing validation, assume the worst-- a data integrity
         // mismatch. If not, these tests won't be performed, and we can assume
         // the best.
-        let failed = crc32c || md5;
-
-        if (validateStream) {
+        // We must check if the server decompressed the data on serve because hash
+        // validation is not possible in this case.
+        let failed = (crc32c || md5) && safeToValidate;
```

Comment:
I think i got confused with `failed` variable; what does it do?

Reply:
failed defaults to true (if `crc32c || md5`) and then later gets changed to `false` if it passes. (see below lines.) It's not true that it failed if it wasnt `safeToValidate` to begin with, so we need to make sure that's `true` before assuming `failed`